portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -17,52 +17,77 @@
 import com.google.appengine.api.datastore.DatastoreService;
 import com.google.appengine.api.datastore.DatastoreServiceFactory;
 import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
 import com.google.gson.Gson;
 import java.io.IOException;
 import java.util.Arrays; 
 import java.util.ArrayList;
+import java.util.List;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
 
-
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
+/** Servlet that returns comment content. */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
-  ArrayList<String> commentList = new ArrayList<String>();
-
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    // Convert comment_list into a JSON object.
+    // Retreive comments from database.
+    Query query = new Query("Comment").addSort("text", SortDirection.ASCENDING);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    // Limit number of comments and save in  comments array.
+    List<String> comments = new ArrayList<>();
+    int counter = 0;
+    int commentLimit = getCommentLimit(request);
+
+    for (Entity entity: results.asIterable()) {
+      String text = (String) entity.getProperty("text");
+      comments.add(text);
+      counter++;
```

Comment:
I guess the instruction has not given this information (But I want to let you know since it is helpful). You don't have to do it. You could potentially 
- fetch a fixed/max amount of items from the database. [See](https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/PreparedQuery.html#countentities). This will increase the efficiency a lot without keeping a counter inside the iteration. For example, if the database has 1K items and you just need 10 of them. A full lookup of 1000 items will cost 100 times more than a look up with 10 items back, and you don't have to count yourself afterwords.
Here is an example:
```
import com.google.appengine.api.datastore.FetchOptions;

 Cursor cursor = ...

 ...

 // limit 10
 datastoreService.prepare(query).asList(FetchOptions.Builder.withLimit(10));

 // limit 10, start cursor
 datastoreService.prepare(query).asList(FetchOptions.Builder.withLimit(10).startCursor(cursor));
 ```
- get the result as a list but not only iterable, and there are other options as well. [See](https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/PreparedQuery#aslist).

Reply:
ooohhh nice thank you